src/hooks/useImmediateState.ts

```diff
@@ -0,0 +1,32 @@
+import { useRef, useEffect, useState, useCallback } from 'react';
+
+type SetStateAction<S> = S | ((prevState: S) => S);
+type SetState<S> = (action: SetStateAction<S>) => void;
+
+export const useImmediateState = <S extends {}>(init: S): [S, SetState<S>] => {
+  const isMounted = useRef(false);
+  const initialState = useRef<S>({ ...init });
+  const [state, setState] = useState<S>(initialState.current);
+
+  const updateState: SetState<S> = useCallback((action: SetStateAction<S>) => {
+    // If the component is currently before its initial mount, update
+    // the state immediately
+    if (isMounted.current) {
+      setState(action);
```

Comment:
so we set the potential function as state? Sorry if this question is weird trying to grasp it

Reply:
If we're in "will mount" basically we don't call setState and update the state mutably

Reply:
Oh, yes good point thanks for clarifying.